### PR TITLE
Cleanup char tests

### PR DESCRIPTION
**Pull request by hughbe**
- Take advantage of xunit
- Split up some tests to ensure tests only test a single thing

Contributes to #6286
Cherry picked from #5490

### CODE REVIEW

**Comment by hughbe** [comment]
Test Innerloop OSX Debug Build and Test
Test Innerloop Windows_NT Debug Build and Test

File: src/System.Runtime/tests/System/Char.cs [hunk 1]
```diff
@@ -5,516 +5,502 @@
...
+    {
+        VerifyConvertFromUtf32(0x10000, "\uD800\uDC00");
+        VerifyConvertFromUtf32(0x103FF, "\uD800\uDFFF");
+        VerifyConvertFromUtf32(0xFFFFF, "\uDBBF\uDFFF");
+        VerifyConvertFromUtf32(0x10FC00, "\uDBFF\uDC00");
+        VerifyConvertFromUtf32(0x10FFFF, "\uDBFF\uDFFF");
+        VerifyConvertFromUtf32(0, "\0");
+        VerifyConvertFromUtf32(0x3FF, "\u03FF");
+        VerifyConvertFromUtf32(0xE000, "\uE000");
+        VerifyConvertFromUtf32(0xFFFF, "\uFFFF");
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/Char.cs]: Why is this not a theory?  Is it due to the xunit issue around serializing string arguments with characters invalid in XML?

File: src/System.Runtime/tests/System/Char.cs [hunk 2]
```diff
@@ -5,516 +5,502 @@
...
-        ValidateConvertFromUtf32(0x10FC00, "\uDBFF\uDC00");
-        ValidateConvertFromUtf32(0x10FFFF, "\uDBFF\uDFFF");
-        ValidateConvertFromUtf32(0, "\0");
-        ValidateConvertFromUtf32(0x3FF, "\u03FF");
-        ValidateConvertFromUtf32(0xE000, "\uE000");
-        ValidateConvertFromUtf32(0xFFFF, "\uFFFF");
-        ValidateConvertFromUtf32(0xD800, null);
-        ValidateConvertFromUtf32(0xDC00, null);
-        ValidateConvertFromUtf32(0xDFFF, null);
-        ValidateConvertFromUtf32(0x110000, null);
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/Char.cs]: What happened to this test case?

File: src/System.Runtime/tests/System/Char.cs [hunk 3]
```diff
@@ -5,516 +5,502 @@
...
-    public static void TestConvertToUtf32StrInt()
-    {
-        // Int32 Char.ConvertToUtf32(String, Int32)
-        ValidateconverToUtf32<Exception>("\uD800\uDC00", 0, 0x10000);
-        ValidateconverToUtf32<Exception>("\uD800\uD800\uDFFF", 1, 0x103FF);
-        ValidateconverToUtf32<Exception>("\uDBBF\uDFFF", 0, 0xFFFFF);
-        ValidateconverToUtf32<Exception>("\uDBFF\uDC00", 0, 0x10FC00);
-        ValidateconverToUtf32<Exception>("\uDBFF\uDFFF", 0, 0x10FFFF);
+    public static void TestConvertToUtf32_String_Int()
+    {
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/Char.cs]: Same [Theory] question/comment.

File: src/System.Runtime/tests/System/Char.cs (hunk 1 above)
Comment (hughbe) [inline src/System.Runtime/tests/System/Char.cs]: Correct.

File: src/System.Runtime/tests/System/Char.cs [hunk 4]
```diff
@@ -5,516 +5,502 @@
...
+
+        Assert.Throws<ArgumentException>("s", () => char.ConvertToUtf32("\uD800\uD800", 0)); // High, high
+        Assert.Throws<ArgumentException>("s", () => char.ConvertToUtf32("\uD800\uD7FF", 0)); // High, non-surrogate
+        Assert.Throws<ArgumentException>("s", () => char.ConvertToUtf32("\uD800\u0000", 0)); // High, non-surrogate
+        Assert.Throws<ArgumentException>("s", () => char.ConvertToUtf32("\uD800\uD800", 1)); // High, high
+
+        Assert.Throws<ArgumentException>("s", () => char.ConvertToUtf32("\uDC01\uD940", 0)); // Low, high
+        Assert.Throws<ArgumentException>("s", () => char.ConvertToUtf32("\uDD00\uDE00", 0)); // Low, low
+        Assert.Throws<ArgumentException>("s", () => char.ConvertToUtf32("\uDF01\u0000", 0)); // Low, non-surrogate
+        Assert.Throws<ArgumentException>("s", () => char.ConvertToUtf32("\uDC01\uD940", 1)); // Low, high
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/Char.cs]: Nit: in the "\uD800\u0000" case, you reorganized the lines so that both cases of it were next to each other, but here you didn't.  It'd be good to be consistent.

File: src/System.Runtime/tests/System/Char.cs [hunk 5]
```diff
@@ -5,516 +5,502 @@
...
+        Assert.Throws<ArgumentException>("s", () => char.ConvertToUtf32("\uDD00\uDE00", 1)); // Low, high
+
+        Assert.Throws<ArgumentOutOfRangeException>("index", () => char.ConvertToUtf32("abcde", -1)); // Index < 0
+        Assert.Throws<ArgumentOutOfRangeException>("index", () => char.ConvertToUtf32("abcde", 5)); // Index >= string.Length
+        Assert.Throws<ArgumentOutOfRangeException>("index", () => char.ConvertToUtf32("", 0)); // Index >= string.Length
     }
 
     [Fact]
-    public static void TestEquals()
+    public static void TestConvertToUtf32_Char_Char()
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/Char.cs]: Same comment about adding a comment with the issue number.

File: src/System.Runtime/tests/System/Char.cs [hunk 6]
```diff
@@ -748,53 +733,53 @@ public static void TestIsWhiteSpace()
...
+        foreach (char c in GetTestChars(categories))
             Assert.True(char.IsWhiteSpace(c.ToString(), 0));
 
         // Some control chars are also considered whitespace for legacy reasons.
-        //if ((c >= '\x0009' && c <= '\x000d') || c == '\x0085')
-        Assert.True(char.IsWhiteSpace('\u000b'.ToString(), 0));
-        Assert.True(char.IsWhiteSpace('\u0085'.ToString(), 0));
+        // if ((c >= '\x0009' && c <= '\x000d') || c == '\x0085')
+        Assert.True(char.IsWhiteSpace('\u000b'));
+        Assert.True(char.IsWhiteSpace('\u0085'));
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/Char.cs]: Why did you change these?  You're now using the IsWhiteSpace(char) overload, whereas this test is about IsWhiteSpace(string, int).

File: src/System.Runtime/tests/System/Char.cs [hunk 7]
```diff
@@ -894,44 +866,58 @@ public static void TestToUpperInvariant()
...
-        ValidateTryParse("", '\0', false);
-        ValidateTryParse("\n\r", '\0', false);
-        ValidateTryParse("kj", '\0', false);
-        ValidateTryParse(" a", '\0', false);
-        ValidateTryParse("a ", '\0', false);
-        ValidateTryParse("\\u0135", '\0', false);
-        ValidateTryParse("\u01356", '\0', false);
-        ValidateTryParse("\ud801\udc01", '\0', false);   // surrogate pair
+    public static void TestParse_Surrogate()
+    {
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/Char.cs]: Why are these separated out from TestParse?  Again because of xunit?  If so, as noted please add a comment with the issue number so these can be recombined later.

**Comment by stephentoub** [comment]
Thanks, @hughbe. Aside from my comments, LGTM.

**Comment by hughbe** [comment]
Thanks for the comments. Issue #7166 does frustrate the cleanup a little, but I'm hoping its fixed soon

**Comment by stephentoub** [comment]
Thanks.
